src/msg/async/ProtocolV2.cc

```diff
@@ -97,10 +97,17 @@ ProtocolV2::ProtocolV2(AsyncConnection *connection)
       tx_frame_asm(&session_stream_handlers, false),
       rx_frame_asm(&session_stream_handlers, false),
       next_tag(static_cast<Tag>(0)),
-      keepalive(false) {
+      keepalive(false),
+      th_timer(cct, th_limit_lock, false) {
+  th_timer.init();
+  th_message_callback = new LambdaContext([this]() {
```

Comment:
I thought I would work with [time event](https://github.com/ceph/ceph/blob/757ff2922909cfc60add4509587e6345585caffd/src/msg/async/ProtocolV2.cc#L1525) originally; to display the debug message once in 30s. But it seems to be associated with the connection. Then the usage of SafeTimer in OSD.cc looks good/independent approach to me. If you have a better idea, please share. Thanks.

Reply:
I think it should be as simple as saving a `mono_time` when you print the debug message. Then check if it's been more than 30s since the last message every time you output the debug message?

Reply:
The original patch that I had sent via email was similar. We could save a `mono_time` in `ProtocolV2::throttle_message()`, `ProtocolV2::throttle_bytes()` and `ProtocolV2::throttle_dispatch_queue()` when the throttle hits the limit and then check for 30s when we output the debug message. But the question/unknown part for me is: where do we output the debug message? If it is inside `ProtocolV2::read_event()` then how can I ensure that the `read_event()` would be called within 30s or too frequent? (any idea, please share). That's where I think a callback mechanism would be appropriate.

Reply:
It is not appropriate to use a timer to limit logs. You can simply compare the interval between two logs before and after the log is generated. If the interval is less than 30 seconds, no log is generated. User can run the "ceph daemon osd.$i perf dump " command to view the number of failed attempts.